rfcs/0148-pipe-operator.md

```diff
@@ -0,0 +1,223 @@
+it can directly benefit from `builtins.pipe` and its better error diagnostic by overriding `lib.pipe`.
+Elevating a Nixpkgs library function to a builtin has been done several times before,
+for example `bitAnd`, `splitVersion` and `concatStringsSep`.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Tooling support
+
+Like any language extension, this will require the available Nix tooling to be updated.
+Updating parsers should be pretty easy, as the syntax changes to the language are fairly minimal.
+Tooling that evaluates Nix code in some way or does static code analysis should be easy to support too,
+since one may treat the operator as syntactic sugar for function application.
+No fundamentally new semantics are introduced to the language.
+
+# Prior art
+
+Nickel has `|>` too, with the same name and semantics.
+
+F# has `|>`, called "pipe-forward" operator, with the same semantics.
+Additionally, it also has "pipe-backward" `<|` and `>>`/`<<` for forwards and backwards function composition.
+`<|` is equivalent to function application, however its lower binding order allows removing parentheses:
+`g (f a)` = `g <| f a`. All these operators have the same precedence and are left-associative.
+
+Elm has the same operators as F#.
+
+Haskell has the (backwards) function composition operator `.` in its prelude: `(g . f) a` = `g (f a)`.
+It also has "reverse application" `&`, which is roughly equivalent to `|>`,
+and `$`, which is function application again but right-associative and very weakly binding.
+`.` binds stronger than both.
+
+`|>` is definable as an infix function in several other programming languages,
+and in even more languages as macro or higher-order function (including Nix, that's `lib.pipe`).
+
+# Alternatives
+[alternatives]: #alternatives
+
+For each change this RFC proposes, there is always the trivial alternative of not doing it. See #drawbacks.
+
+## More operators
+
+We could use the occasion and introduce more operators like those mentioned above.
+
+Function composition is mostly interesting for the so-called "point-free" programming style,
+where partially applied compositions of functions are preferred over the introduction of lambda terms.
+However, Nix is not well suited for that programming style for various reasons,
+nor would that point-free style have nearly as many applications in real-world Nixpkgs code.
+
+F#'s reverse-pipe operator has a lot less use due to its left-associativity,
+but a right-associative version of it more similar to Haskell's `$` might be an alternative:
+
+```nix
+defaultPrefsFile = 
+  pkgs.writeText "nixos-default-prefs.js" <|
+  lib.concatStringsSep "\n" <|
+  lib.mapAttrsToList (
+    key: value: ''
+      // ${value.reason}
+      pref("${key}", ${builtins.toJSON value.value});
+    ''
+  )
+  defaultPrefs
+  ;
+```
+
+Adding both pipe directions raises questions about how these two interact when used together.
```

Comment:
Interaction without parentheses could be forbidden, but this is a kind of thing we do yet in the language, so it has some small extra implementation cost that has to be repeated in all serious implementations (e.g. `nix`, `hnix`, `tvix`, linters that care)

Reply:
Are there any precedents in other languages for forbidding to mix certain operators like this? (On a syntax level, ignoring types) How would one formalize such a grammar?

Reply:
Just tried this in Haskell:
```
ghci> import Flow
ghci> id <| 42 |> id

<interactive>:2:1: error:
    Precedence parsing error
        cannot mix ‘<|’ [infixr 0] and ‘|>’ [infixl 0] in the same infix expression

Reply:
Turns out Scala has a similar restriction too.